File: week_2/code/day_8/e_task_scheduler.py
import heapq
from collections import Counter, deque

class Solution:
    def leastInterval(self, tasks: list[str], n: int) -> int:
        # Count task frequencies
        count = Counter(tasks)
        
        # Max-heap to store frequencies (use negative values for heapq)
        max_heap = [-cnt for cnt in count.values()]
        heapq.heapify(max_heap)
        
        time = 0
        # Queue to store the cooldown: pairs of [remaining_count, unlock_time]
        q = deque()
        
        while max_heap or q:
            time += 1
            
            # If we have tasks to process
            if max_heap:
                # Process the most frequent task
                cnt = 1 + heapq.heappop(max_heap) # +1 because we are dealing with negative numbers
                
                # If there are still instances of this task left
                if cnt != 0:
                    # It will be ready at time + n
                    q.append([cnt, time + n])
            
            # If the task at front of the queue is ready
            if q and q[0][1] == time:
                # Push back into max heap
                heapq.heappush(max_heap, q.popleft()[0])
                
        return time

# Example Usage
if __name__ == "__main__":
    sol = Solution()
    print(sol.leastInterval(["A","A","A","B","B","B"], 2)) # 8
    print(sol.leastInterval(["A","C","A","B","D","B"], 1)) # 6
    print(sol.leastInterval(["A","A","A","B","B","B"], 3)) # 10

File: week_2/code/day_8/e_task_scheduler.js
const MinHeap = require('./min_heap');

/**
 * @param {character[]} tasks
 * @param {number} n
 * @return {number}
 */
var leastInterval = function (tasks, n) {
    // Count task frequencies
    const count = new Map();
    for (let task of tasks) {
        count.set(task, (count.get(task) || 0) + 1);
    }

    // Max-heap to store frequencies
    const maxHeap = new MinHeap((a, b) => b - a);

    for (let val of count.values()) {
        maxHeap.push(val);
    }

    let time = 0;
    // Queue to store the cooldown: { remainingCount, unlockTime }
    const q = [];

    while (maxHeap.size() > 0 || q.length > 0) {
        time++;

        if (maxHeap.size() > 0) {
            let cnt = maxHeap.pop(); // Most frequent remaining
            cnt--; // Process one instance

            if (cnt > 0) {
                // Add to cooldown queue
                q.push({ remainingCount: cnt, unlockTime: time + n });
            }
        }

        // Check if any task in cooldown is ready
        if (q.length > 0 && q[0].unlockTime === time) {
            maxHeap.push(q.shift().remainingCount);
        }
    }

    return time;
};

// Example Usage
console.log(leastInterval(["A", "A", "A", "B", "B", "B"], 2)); // 8
console.log(leastInterval(["A", "C", "A", "B", "D", "B"], 1)); // 6
console.log(leastInterval(["A", "A", "A", "B", "B", "B"], 3)); // 10
